### PR TITLE
PIM-7359: add customer volumes on new axes in SystemInfo

### PR DESCRIPTION
**Pull request by Doodoune**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: tests/features/volume-monitoring/systeminfo.feature [hunk 1]
```diff
@@ -0,0 +1,83 @@
+Feature: System Info
+  In order to guarantee the performance of the PIM
+  As an administrator user
+  I want to monitor volumes in system information
+
+  @acceptance-back
+  Scenario: Monitor the number of channels
```
Comment (ahocquard) [inline tests/features/volume-monitoring/systeminfo.feature]: My proposition, feel free to challenge it:
```
  Scenario: Gather customers statistics about the number of channels
    Given a catalog with 3 channels for the customer Foo
    When statistics are collected from the customer's PIM
    Then Akeneo statistics engine store a number of 3 channels for the customer Foo
```

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+                      WHERE lft < node.lft
+                            AND rgt > node.rgt
+                            AND node.lft <> 0
+                    )
+                      AS depth
+                  FROM `pim_catalog_category` AS node
+                  WHERE lft <> 0
+                  ORDER BY lft
+            )as max_depth;
+SQL;
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php]:
```    
SELECT 	
    	MAX(lvl) as max,
    	CEIL(AVG(lvl)) as average 
    FROM pim_catalog_category;
```

(your request was wrong a on bigger dataset btw, and not taking into account the root tree)

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/CategoryInOneCategoryContext.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     */
+    public function theReportReturnsThatTheMaximumOfCategoryInOneCategoryIs(int $maxOfCategoryInOneCategory): void
+    {
+        $volumes = $this->reportContext->getVolumes();
+
+        Assert::eq($maxOfCategoryInOneCategory, $volumes['average_max_category_in_one_category']['value']['max']);
+    }
+
+    /**
+     * @Then the report returns that the average of category in one category is :avgOfCategoryInOneCategory
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/CategoryInOneCategoryContext.php]: Fix

File: tests/features/volume-monitoring/systeminfo.feature [hunk 4]
```diff
@@ -0,0 +1,83 @@
+Feature: System Info
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/features/volume-monitoring/systeminfo.feature]: Fix

File: tests/features/volume-monitoring/systeminfo.feature [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  In order to guarantee the performance of the PIM
+  As an administrator user
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/features/volume-monitoring/systeminfo.feature]: Fix

File: tests/features/volume-monitoring/systeminfo.feature [hunk 6]
```diff
@@ -0,0 +1,83 @@
+Feature: System Info
+  In order to guarantee the performance of the PIM
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply tests/features/volume-monitoring/systeminfo.feature]: Fix

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/QueryTestCase.php [hunk 7]
```diff
@@ -142,6 +143,62 @@ protected function createCategory(array $data = []) : CategoryInterface
         return $category;
     }
 
+    /**
+     * @param int $numberOfSubCategory
+     * @return CategoryInterface
+     */
+    protected function createCategoryWithSubCategories(int $numberOfSubCategory) : CategoryInterface
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/QueryTestCase.php]: You're right, I move it into AverageMaxCategoriesInOneCategoryIntegration

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/QueryTestCase.php [hunk 8]
```diff
@@ -190,6 +247,21 @@ protected function createFamilyVariant(array $data = []) : FamilyVariantInterfac
         return $family;
     }
 
+    /**
+     * @param array $data
+     * @return UserInterface
+     */
+    protected function createUser(array $data = []) : UserInterface
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/QueryTestCase.php]: Yep, I move it into CountUsersIntegration

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/Query/CountUsersIntegration.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     * @param int $numberOfUsers
+     */
+    protected function createUsers(int $numberOfUsers)
+    {
+        $i = 0;
+        while ($i < $numberOfUsers) {
+            $this->createUser([
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/Query/CountUsersIntegration.php]: corrected

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/Query/CountUsersIntegration.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Pim\Bundle\CatalogVolumeMonitoringBundle\tests\Integration\Persistence\Query;
+
+use PHPUnit\Framework\Assert;
+use Pim\Bundle\CatalogVolumeMonitoringBundle\tests\Integration\Persistence\QueryTestCase;
+use Pim\Bundle\UserBundle\Entity\UserInterface;
+
+class CountUsersIntegration extends QueryTestCase
+{
+    public function testGetCountOfUsers()
```
[the remark this answers is not in the document]
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Integration/Persistence/Query/CountUsersIntegration.php]: Fix

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php (hunk 2 above)
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php]: You're right, fixed

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoriesInOneCategory.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+              FROM (
+                 SELECT COUNT(c.parent_id) as count_parent_id
+                  FROM pim_catalog_category c
+                 WHERE c.parent_id is not null
+                  GROUP BY c.parent_id
+             ) as count_parent;
+SQL;
+        $result = $this->connection->query($sql)->fetch();
+
+        $volume = new AverageMaxVolumes((int) $result['max'], (int) $result['average'], $this->limit, self::VOLUME_NAME);
```
Comment (AymericPlanche) [inline src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoriesInOneCategory.php]: is there always a result ? sanity check before access a array key ;)

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function fetch(): AverageMaxVolumes
+    {
+        $sql = <<<SQL
+        SELECT 	
+    	MAX(lvl) as max,
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php]: care about indentation, it helps to read requests

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Resources/config/queries.yml [hunk 13]
```diff
@@ -1,14 +1,14 @@
 services:
     pim_volume_monitoring.persistence.query.average_max_attributes_per_family:
-        class: 'Pim\Bundle\CatalogVolumeMonitoringBundle\Persistence\Query\Sql\AverageMaxAttributesPerFamily'
+        class: Pim\Bundle\CatalogVolumeMonitoringBundle\Persistence\Query\Sql\AverageMaxAttributesPerFamily
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogVolumeMonitoringBundle/Resources/config/queries.yml]: put it in parameters and follow the convention please. Also, string should always be surrend by suote in YAML AFAIK (but it supports to not have it).

Please fix the other services.

File: tests/features/volume-monitoring/system_info.feature [hunk 14]
```diff
@@ -0,0 +1,83 @@
+Feature: System Info
+  In order to gather statistics to better know the usage of the PIM
```
Comment (ahocquard) [inline tests/features/volume-monitoring/system_info.feature]: this is not system info anymore.
It is volume_customer_statistics.
```
Feature: Volume statistics of the customers

In order to better know the use of the PIM
As Akeneo Company
I want to monitor anonymously volumes of the customers
```

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/UserContext.php [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $volumes = $this->reportContext->getVolumes();
+
+        Assert::true($volumes['count_users']['has_warning']);
+    }
+
+    /**
+     * @Then the report does not warn the users that the number of users is high
+     */
+    public function theReportDoesNotWarnTheUsersThatTheNumberIsHigh(): void
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/UserContext.php]: Seems to be not used. Same for other context: delete useless function

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoriesInOneCategory.php (hunk 11 above)
Comment (ahocquard) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoriesInOneCategory.php]: It always return a result.
As it used typecasting, if it's null, it equals 0.

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/UserContext.php (hunk 15 above)
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/tests/Acceptance/Context/UserContext.php]: fix

File: tests/features/volume-monitoring/system_info.feature (hunk 14 above)
Comment (Doodoune) [reply tests/features/volume-monitoring/system_info.feature]: fix

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Resources/config/queries.yml (hunk 13 above)
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/Resources/config/queries.yml]: fix

File: src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php (hunk 12 above)
Comment (Doodoune) [reply src/Pim/Bundle/CatalogVolumeMonitoringBundle/Persistence/Query/Sql/AverageMaxCategoryLevels.php]: fix

File: tests/features/volume-monitoring/volume_customer_statistics.feature [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+  @acceptance-back
+  Scenario: Gather customers statistics about the number of category trees
+    Given a catalog with 7 category trees
+    When statistics of the customer's catalog are collected
+    Then Akeneo statistics engine stores a number of 7 category trees for this customer
+
+  @acceptance-back
+  Scenario: Gather customers statistics about the maximum of categories in one category
+    Given a catalog with 8 category in one category
```
Comment (NolwennP) [inline tests/features/volume-monitoring/volume_customer_statistics.feature]: 8 categories

**Pull request merged**